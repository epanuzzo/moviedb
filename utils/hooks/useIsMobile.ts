import { useEffect, useState } from "react";
import { tabletBreakpoint } from "@/utils/config";

const useIsMobile = () => {
  const [isMobile, setIsMobile] = useState<boolean>(false);

  useEffect(() => {
    const handleResize = () => {
      setIsMobile(window.innerWidth <= tabletBreakpoint);
    };

    handleResize();
    window.addEventListener("resize", handleResize);
    return () => {
      window.removeEventListener("resize", handleResize);
    };
  }, []);

  return isMobile;
};

export default useIsMobile;
