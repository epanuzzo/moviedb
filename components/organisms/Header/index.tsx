import MenuIcon from "@/components/atoms/MenuIcon";
import MenuItems from "@/components/molecules/MenuItems";
import SearchBar from "@/components/molecules/SearchBar";
import React, { useCallback, useState } from "react";
import { TbMovie } from "react-icons/tb";
import useIsMobile from "@/utils/hooks/useIsMobile";
import useActivePage from "@/utils/hooks/useActivePage";
import { slogan } from "@/utils/config";
import { AppDispatch } from "@/redux/store";
import { useDispatch } from "react-redux";
import { debounce } from "@/utils/helpers/debounce";
import { searchMovieByName } from "@/redux/slices/movieSlice";

const Header: React.FC = () => {
  const dispatch = useDispatch<AppDispatch>();
  const [searchText, setSearchText] = useState("");
  const [isMenuOpen, setIsMenuOpen] = useState(false);

  const isMobile = useIsMobile();
  const { activePage } = useActivePage();

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const handleSearchChange = useCallback(
    debounce((e: unknown) => {
      const event = e as React.ChangeEvent<HTMLInputElement>;
      setSearchText(event.target.value);
    }, 300),
    []
  );

  const handleSearchSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    dispatch(searchMovieByName(searchText));
  };

  const toggleMenu = () => {
    setIsMenuOpen(!isMenuOpen);
  };

  return (
    <header className="bg-gray-900 text-white py-4 px-8 border-solid border-b border-slate-300">
      <div className="flex justify-between items-center">
        <div className="flex items-center space-x-2">
          <TbMovie className="w-7 h-7 text-white opacity-30" />
          <span className="text-lg font-semibold opacity-30">{slogan}</span>
          {!isMobile && <MenuItems active={activePage} />}
        </div>
        {!isMobile && (
          <SearchBar
            handleSearchSubmit={handleSearchSubmit}
            handleSearchChange={handleSearchChange}
          />
        )}
        {isMobile && <MenuIcon toggleMenu={toggleMenu} />}
      </div>
      {isMobile && (
        <div className="mt-4">
          <MenuItems
            active={activePage}
            isMobile
            wrapperClassName={`${isMenuOpen ? "block mb-4" : "hidden"}`}
          />
          <SearchBar
            handleSearchSubmit={handleSearchSubmit}
            handleSearchChange={handleSearchChange}
          />
        </div>
      )}
    </header>
  );
};

export default Header;
