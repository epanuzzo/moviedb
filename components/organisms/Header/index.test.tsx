import React from "react";
import renderer from "react-test-renderer";
import Header from "./index";
import { useDispatch } from "react-redux";
import { useRouter } from "next/router";
import useIsMobile from "@/utils/hooks/useIsMobile";

jest.mock("react-redux", () => ({
  useDispatch: jest.fn(),
}));

jest.mock("next/router", () => ({
  useRouter: jest.fn(),
}));

jest.mock("@/utils/hooks/useIsMobile", () => jest.fn());

describe("Header", () => {
  let mockDispatch: jest.Mock;
  let mockRouter: unknown;

  beforeEach(() => {
    mockDispatch = jest.fn();
    (useDispatch as unknown as jest.Mock).mockReturnValue(mockDispatch);
    mockRouter = {
      pathname: "/current-page",
      push: jest.fn(),
      query: { id: "1" },
    };
    (useRouter as jest.Mock).mockReturnValue(mockRouter);
  });

  it("renders component", () => {
    const tree = renderer.create(<Header />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders component with alt props", () => {
    (useIsMobile as jest.Mock).mockReturnValue(true);
    const tree = renderer.create(<Header />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
