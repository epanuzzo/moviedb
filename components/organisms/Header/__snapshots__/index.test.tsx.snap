// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders component 1`] = `
<header
  className="bg-gray-900 text-white py-4 px-8 border-solid border-b border-slate-300"
>
  <div
    className="flex justify-between items-center"
  >
    <div
      className="flex items-center space-x-2"
    >
      <svg
        className="w-7 h-7 text-white opacity-30"
        fill="none"
        height="1em"
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
        style={
          {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M4 4m0 2a2 2 0 0 1 2 -2h12a2 2 0 0 1 2 2v12a2 2 0 0 1 -2 2h-12a2 2 0 0 1 -2 -2z"
        />
        <path
          d="M8 4l0 16"
        />
        <path
          d="M16 4l0 16"
        />
        <path
          d="M4 8l4 0"
        />
        <path
          d="M4 16l4 0"
        />
        <path
          d="M4 12l16 0"
        />
        <path
          d="M16 8l4 0"
        />
        <path
          d="M16 16l4 0"
        />
      </svg>
      <span
        className="text-lg font-semibold opacity-30"
      >
        Best of '90 Movies
      </span>
      <nav
        className="hidden md:flex space-x-4 pl-4 "
      >
        <a
          className="text-1xl "
          href="/"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
        >
          Home
        </a>
        <a
          className="text-1xl "
          href="/about"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
        >
          About
        </a>
      </nav>
    </div>
    <form
      className="relative "
      onSubmit={[Function]}
    >
      <input
        className="w-full bg-transparent placeholder:text-slate-400 text-slate-700 text-sm border border-slate-200 rounded-md pl-3 pr-10 py-2 transition duration-300 ease focus:outline-none focus:border-slate-400 hover:border-slate-300 shadow-sm focus:shadow"
        onChange={[Function]}
        placeholder="Enter your text"
        type="text"
      />
      <button
        className="absolute right-1 top-1 rounded bg-slate-800 p-1.5 border border-transparent text-center text-sm text-white transition-all shadow-sm hover:shadow focus:bg-slate-700 focus:shadow-none active:bg-slate-700 hover:bg-slate-700 active:shadow-none disabled:pointer-events-none disabled:opacity-50 disabled:shadow-none"
        type="submit"
      >
        <svg
          className="w-4 h-4"
          fill="currentColor"
          viewBox="0 0 16 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M9.965 11.026a5 5 0 1 1 1.06-1.06l2.755 2.754a.75.75 0 1 1-1.06 1.06l-2.755-2.754ZM10.5 7a3.5 3.5 0 1 1-7 0 3.5 3.5 0 0 1 7 0Z"
            fillRule="evenodd"
          />
        </svg>
      </button>
    </form>
  </div>
</header>
`;

exports[`Header renders component with alt props 1`] = `
<header
  className="bg-gray-900 text-white py-4 px-8 border-solid border-b border-slate-300"
>
  <div
    className="flex justify-between items-center"
  >
    <div
      className="flex items-center space-x-2"
    >
      <svg
        className="w-7 h-7 text-white opacity-30"
        fill="none"
        height="1em"
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
        style={
          {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M4 4m0 2a2 2 0 0 1 2 -2h12a2 2 0 0 1 2 2v12a2 2 0 0 1 -2 2h-12a2 2 0 0 1 -2 -2z"
        />
        <path
          d="M8 4l0 16"
        />
        <path
          d="M16 4l0 16"
        />
        <path
          d="M4 8l4 0"
        />
        <path
          d="M4 16l4 0"
        />
        <path
          d="M4 12l16 0"
        />
        <path
          d="M16 8l4 0"
        />
        <path
          d="M16 16l4 0"
        />
      </svg>
      <span
        className="text-lg font-semibold opacity-30"
      >
        Best of '90 Movies
      </span>
    </div>
    <button
      className="ml-2 md:hidden text-white"
      onClick={[Function]}
    >
      <svg
        className="w-6 h-6"
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M3 4h14a1 1 0 0 1 0 2H3a1 1 0 0 1 0-2zm0 5h14a1 1 0 0 1 0 2H3a1 1 0 0 1 0-2zm0 5h14a1 1 0 0 1 0 2H3a1 1 0 0 1 0-2z"
          fillRule="evenodd"
        />
      </svg>
    </button>
  </div>
  <div
    className="mt-4"
  >
    <nav
      className="flex md:hidden flex-col space-y-4 pt-4 hidden"
    >
      <a
        className="text-1xl "
        href="/"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
      >
        Home
      </a>
      <a
        className="text-1xl "
        href="/about"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
      >
        About
      </a>
    </nav>
    <form
      className="relative "
      onSubmit={[Function]}
    >
      <input
        className="w-full bg-transparent placeholder:text-slate-400 text-slate-700 text-sm border border-slate-200 rounded-md pl-3 pr-10 py-2 transition duration-300 ease focus:outline-none focus:border-slate-400 hover:border-slate-300 shadow-sm focus:shadow"
        onChange={[Function]}
        placeholder="Enter your text"
        type="text"
      />
      <button
        className="absolute right-1 top-1 rounded bg-slate-800 p-1.5 border border-transparent text-center text-sm text-white transition-all shadow-sm hover:shadow focus:bg-slate-700 focus:shadow-none active:bg-slate-700 hover:bg-slate-700 active:shadow-none disabled:pointer-events-none disabled:opacity-50 disabled:shadow-none"
        type="submit"
      >
        <svg
          className="w-4 h-4"
          fill="currentColor"
          viewBox="0 0 16 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M9.965 11.026a5 5 0 1 1 1.06-1.06l2.755 2.754a.75.75 0 1 1-1.06 1.06l-2.755-2.754ZM10.5 7a3.5 3.5 0 1 1-7 0 3.5 3.5 0 0 1 7 0Z"
            fillRule="evenodd"
          />
        </svg>
      </button>
    </form>
  </div>
</header>
`;
